### PR TITLE
Vendored FLA: re-resolve kernel closures on modules imported before us

### PR DESCRIPTION
**Pull request by danielhanchen**
### Summary

A `qwen3_5` modeling module imported before `patch_vendor_fla()` runs never uses the fla kernels at all, even though everything reports that fla is available.

Since huggingface/transformers#47630 the gated delta kernels are bound by `use_kernel_func_from_hub_with_fallback`, which resolves the implementation inside `decorator()` and closes over it. That happens when the modeling module is imported. If fla is not live at that moment the pure-PyTorch fallback is frozen into the closure, and nothing done to `fla` afterwards can reach it.

`_repair_already_imported_modeling` cannot fix this: it rebinds module globals, and post-#47630 `modeling_qwen3_5` no longer has them. `chunk_gated_delta_rule` and `recurrent_gated_delta_rule` are not module attributes there at all; the live objects are `torch_chunk_gated_delta_rule` and `torch_recurrent_gated_delta_rule`.

This predates the decode alias in #1101 and is not caused by it. The chunked kernel, which is the training path, was already dead on this ordering.

### Measured

Tiny `qwen3_5`, 4 layers, `full_attention_interval = 4`, one forward, counting real entries into `fla.ops.gated_delta_rule.chunk_gated_delta_rule`:

| import order | dispatches to | fla chunk entries |
| --- | --- | --- |
| `patch_vendor_fla` first | `chunk_gated_delta_rule @ fla.ops.gated_delta_rule.chunk` | 3 |
| modeling first, before this PR | `torch_chunk_gated_delta_rule @ transformers...modeling_qwen3_5` | **0** |
| modeling first, with this PR | `chunk_gated_delta_rule @ fla.ops.gated_delta_rule.chunk` | 3 |

Three of four layers are linear attention, so 3 is the expected count. Logits are finite in every case, which is exactly why this is easy to miss: the model works, it is just several times slower and nothing says so.

### Approach

`_repair_kernel_hub_closures` re-applies the decorator to the undecorated function, which `functools.wraps` leaves reachable on `__wrapped__`, once fla is live.

Re-applying rather than writing to `cell_contents` is the substance of the fix, not a style preference. The wrapper closes over the implementation's parameter names as well as the implementation, and it filters `kwargs` down to them. Those lists differ: 10 names for the chunk fallback against 16 for the fla kernel, and 9 against 19 for decode. Swapping only the implementation would leave the fallback's parameter list in place and silently drop arguments the kernel needs.

It runs in the same `finally` as the decode alias and after it, since decode resolves through that alias.

Skips, each deliberate:

- a wrapper already dispatching to something other than its own fallback, so the normal import order is untouched;
- an attribute with no `__wrapped__`, i.e. a transformers predating the decorator;
- a rebuild that still resolves to the fallback, meaning fla genuinely has no such kernel;
- `olmo_hybrid`, by scoping to `_REPAIR_MODELING`. It imports `ShortConvolution`, which is not vendored, so its probe must stay False and binding vendored kernels into it would contradict that.

Failures are caught and logged, like the alias step, so the exit path cannot take `patch_vendor_fla` down.

### Tests

Eight added. The important one asserts the frozen fallback is actually rebuilt; the rest pin the four skips, `_resolved_implementation` on a wrapper with no closure, an absent `hub_kernels`, and that a raising repair still lets `patch_vendor_fla` return.

`tests/test_vendor_fla.py`: 32 passed, 1 failed. The failure is `test_uncovered_model_imports_on_fallback_subprocess`, and I confirmed it fails identically on a clean `main` with this branch stashed, so it is pre-existing drift against transformers 5.15.1.

### CODE REVIEW

File: unsloth_zoo/temporary_patches/fla_vendor.py [hunk 1]
```diff
@@ -1028,6 +1028,82 @@ def _alias_missing_gated_delta_names():
...
+        module = sys.modules.get(f"transformers.models.{package}.modeling_{package}")
+        if module is None:
+            continue
+        for attribute, kernel in _KERNEL_HUB_DECORATED.items():
+            wrapper = getattr(module, attribute, None)
+            original = getattr(wrapper, "__wrapped__", None)
+            if original is None:
+                continue        # not decorated on this transformers
+            if _resolved_implementation(wrapper) is not original:
+                continue        # already dispatching to a real kernel
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/fla_vendor.py]: Correct, and fixed in c9af7440. The staleness case is exactly the one `replaced_real` exists for, and the kernel left in the closure is the miscompiled backward the replacement is meant to avoid, so this was a silent gradient-corruption path on an affected Hopper host.

The check now compares against the kernel fla resolves to *now* rather than against the fallback:

```python
current = _resolved_implementation(wrapper)
live = _live_gated_delta_kernel(kernel)
if live is not None and current is live:
    continue        # already dispatching to the fla that is live now
```

That covers all three states rather than two: a frozen fallback, a purged install's kernel, and a correctly bound live kernel which is left untouched. When the wrapper is stale and there is no live replacement, the rebuild lands on the pure-torch fallback and is taken deliberately, since torch beats calling into an install that is no longer on sys.modules.

Two tests added for the states the old condition got wrong: `test_late_import_repair_replaces_a_purged_install_kernel` and `test_late_import_repair_prefers_torch_over_a_purged_kernel`. The existing live-kernel test now pins `_live_gated_delta_kernel` so it is deterministic rather than depending on whatever fla the test host has.

Also worth noting for the rest of this PR: the kernel-hub tests now skip on a transformers predating #47630, where `_repair_kernel_hub_closures` correctly returns `()` and there is nothing to patch. Suite is 34 passed, 1 failed, and that failure is `test_uncovered_model_imports_on_fallback_subprocess`, which I confirmed fails identically with this branch stashed.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
